vanilla/vanilla
# Fix logic for determining if a Discussion has been read by a user.

This will close https://github.com/vanilla/support/issues/559

We have some logic issues when it comes to determining if a user has actually read all the most recent comments in a discussion. The problem happens when users delete comments. We update the Discussion table's Number of Comments in the Discussion but we do not (and cannot for scaling reasons) update the UserDiscussion table. So if I have read 100 Comments in a Discussion and then some users delete 10 Comments, we will compare the number of Comments I have read with the number of Comments available to be read, we see that I have read 100 comments and there are 90 available. Now as more users add comments it still looks like I have read them all.

This PR:
* Extracts the logic for determining if a user really has read all the comments into a two simple testable methods.
    * One determines the user's correct "Last Viewed" date for that discussion. 
    * The second determines whether the discussion should be marked read or not. If there is a discrepancy between the read status and the number of comments marked read, this will resolve it. However, in the case that the comment count is incommensurable with the read status, it will set unread comments to 0 if the conversation is determined to be read and unread comments to 1 if the conversation is determined to be unread. This last case is not ideal (there may in fact have been more than one unread comment), but it is simpler and more scalable than trying to guess the unread count by checking the dates of the last view against all of the comments' dates. It also may be less confusing to the user than guessing the number.
* Adds some logic to setWatch() in commentmodel to update the userDiscussion table and the discussion object with the correct number of comments read by the user.

## Testing
 - Create a fresh discussion.
 - Add 3 comments.
 - View the 3 comments as a different user.
 - Delete one Comment.
 - Make sure that the Discussion appears to be UnRead when you see it on the Discussions page.
 - Click on the discussion as the different user.
 - Confirm that the CountComments column in the userDiscussion table has a value of 2.
